File: travel-agency/accommodation-management/src/main/java/com/co/unitravel/infrastructure/config/AmqpConfig.java
package com.co.unitravel.infrastructure.config;

import com.co.unitravel.infrastructure.config.constants.AmqpConfigurationConstants;
import org.springframework.amqp.core.AmqpTemplate;
import org.springframework.amqp.core.Binding;
import org.springframework.amqp.core.BindingBuilder;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.core.TopicExchange;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.amqp.support.converter.MessageConverter;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class AmqpConfig {

    @Bean
    public Queue cityAccommodationQueue() {
        return new Queue(AmqpConfigurationConstants.CITY_QUEUE);
    }

    @Bean
    public Binding cityAccommodationBinding(@Qualifier("cityAccommodationQueue") Queue queue, TopicExchange exchange) {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.CITY_ROUTING_KEY);
    }

    @Bean
    public Queue userAccommodationQueue() {
        return new Queue(AmqpConfigurationConstants.USER_QUEUE);
    }

    @Bean
    public Binding userAccommodationBinding(@Qualifier("userAccommodationQueue") Queue queue, TopicExchange exchange) {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.USER_ROUTING_KEY);
    }

    @Bean
    public Queue accommodationAccommodationQueue() {
        return new Queue(AmqpConfigurationConstants.ACCOMMODATION_QUEUE);
    }

    @Bean
    public Binding accommodationAccommodationBinding(@Qualifier("accommodationAccommodationQueue") Queue queue, TopicExchange exchange) {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.ACCOMMODATION_ROUTING_KEY);
    }

    @Bean
    public TopicExchange exchange() {
        return new TopicExchange(AmqpConfigurationConstants.EXCHANGE);
    }


    @Bean
    public MessageConverter converter() {
        return new Jackson2JsonMessageConverter();
    }


    @Bean
    public AmqpTemplate template(ConnectionFactory connectionFactory) {
        final RabbitTemplate rabbitTemplate = new RabbitTemplate(connectionFactory);
        rabbitTemplate.setMessageConverter(converter());
        return rabbitTemplate;
    }
}

File: travel-agency/reservation-management/src/main/kotlin/com/co/unitravel/infrastructure/config/AmqpConfig.kt
package com.co.unitravel.infrastructure.config

import com.co.unitravel.infrastructure.config.constants.AmqpConfigurationConstants
import org.springframework.amqp.core.*
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.annotation.Bean

open class AmqpConfig {

    @Bean
    open fun cityReservationQueue(): Queue {
        return Queue(AmqpConfigurationConstants.CITY_QUEUE)
    }

    @Bean
    open fun cityReservationBinding(@Qualifier("cityReservationQueue") queue: Queue?, exchange: TopicExchange?): Binding {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.CITY_ROUTING_KEY)
    }

    @Bean
    open fun userReservationQueue(): Queue {
        return Queue(AmqpConfigurationConstants.USER_QUEUE)
    }

    @Bean
    open fun userReservationBinding(@Qualifier("userReservationQueue") queue: Queue?, exchange: TopicExchange?): Binding {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.USER_ROUTING_KEY)
    }

    @Bean
    fun accommodationReservationQueue(): Queue {
        return Queue(AmqpConfigurationConstants.ACCOMMODATION_QUEUE)
    }

    @Bean
    fun accommodationReservationBinding(@Qualifier("accommodationReservationQueue") queue: Queue?, exchange: TopicExchange?): Binding {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.ACCOMMODATION_ROUTING_KEY)
    }

    @Bean
    fun accommodationReservationQueuePrice(): Queue {
        return Queue(AmqpConfigurationConstants.ACCOMMODATION_QUEUE_PRICE)
    }

    @Bean
    fun accommodationReservationPriceBinding(@Qualifier("accommodationReservationQueuePrice") queue: Queue?, exchange: TopicExchange?): Binding {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.ACCOMMODATION_PRICE_ROUTING_KEY)
    }

    @Bean
    fun accommodationReservationQueueCapacity(): Queue {
        return Queue(AmqpConfigurationConstants.ACCOMMODATION_QUEUE_CAPACITY)
    }

    @Bean
    fun accommodationReservationCapacityBinding(@Qualifier("accommodationReservationQueueCapacity") queue: Queue?, exchange: TopicExchange?): Binding {
        return BindingBuilder.bind(queue).to(exchange).with(AmqpConfigurationConstants.ACCOMMODATION_CAPACITY_ROUTING_KEY)
    }


    @Bean
    open fun exchange(): TopicExchange {
        return TopicExchange(AmqpConfigurationConstants.EXCHANGE)
    }


    @Bean
    open fun converter(): MessageConverter {
        return Jackson2JsonMessageConverter()
    }


    @Bean
    open fun template(connectionFactory: ConnectionFactory?): AmqpTemplate {
        val rabbitTemplate = RabbitTemplate(connectionFactory)
        rabbitTemplate.messageConverter = converter()
        return rabbitTemplate
    }
}